### PR TITLE
Integrate Atomic Simulation Environment (ASE) as a Job Adapter

### PR DESCRIPTION
**Pull request by kfir4444**
**Summary**

This PR introduces a new job adapter that allows ARC to utilize the Atomic Simulation Environment (ASE) as a backend engine. By integrating ASE, ARC can now run single-point energy calculations, geometry optimizations, and vibrational (frequency) analyses using any ASE-compatible calculator.

  **Key Changes**
   * ASE Adapter Interface: Added the main ASEAdapter class (arc/job/adapters/ase.py) and registered it in the adapter factory (arc/job/adapters/common.py and __init__.py).
   * Standalone Execution Script: Introduced arc/job/adapters/scripts/ase_script.py, a standalone Python script designed to standardise the execution of ASE tasks in a robust, isolated
     environment.
   * Multi-Calculator Support: The script natively supports initializing calculators such as TorchANI (with options for ANI1x, ANI1ccx, ANI2x), xTB, and MOPAC, with fallbacks for
     generic ASE calculators.
   * Flexible Geometry Optimization: Enables geometry optimizations via various engines, including Sella, BFGS, LBFGS, and GPMin. Allows defining constraints (e.g., fixing internal
     coordinates).
   * Robust Vibrational Analysis: Implemented a pure NumPy vibrational analysis function that reliably calculates frequencies, normal modes (Mass Deweighted Normalized), reduced
     masses, and force constants directly from the Hessian, matching physical constants and standards used by TorchANI/ASE.

  **Impact**

  This integration significantly broadens ARC's computational capabilities by tapping directly into the ASE ecosystem, making it trivial to incorporate Machine Learning Potentials (like TorchANI) and any method that interacts with ASE very easily.

### CODE REVIEW

File: arc/job/adapters/scripts/ase_script.py [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+    # Filter out translations and rotations (first 6 modes for non-linear, 5 for linear)
+    # Most ESS only report 3N-6 / 3N-5 modes.
+    # We'll filter modes with very small magnitude if they are in the first 6.
+    # Sorting by magnitude ensures we catch the smallest ones.
+    indices = np.argsort(np.abs(freqs))
+    
+    # Threshold for considering a mode as a translation/rotation (cm^-1)
+    rot_trans_threshold = 10.0
+    
+    num_to_filter = 6 if len(masses) > 2 else 5 if len(masses) == 2 else 0
```
Comment (alongd) [inline arc/job/adapters/scripts/ase_script.py]: This filters 6 modes for any molecule with >2 atoms. A linear polyatomic (like CO2, HCN, ...)
has 3N−5 vibrations, so it should filter only 5.
I think ARC has an `is_linear()`, try implementing?

File: arc/job/adapters/ase_adapter.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    def execute_incore(self) -> None:
+        """
+        Execute the job incore.
+        """
+        self.write_input_file()
+        cmd = [self.python_executable, self.script_path, '--yml_path', self.local_path]
+        process = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
+        if process.returncode != 0:
+            logger.error(f"ASE job failed incore:\n{process.stderr}")
+        self.parse_results()
```
Comment (alongd) [inline arc/job/adapters/ase_adapter.py:174]: On failure it logs but still calls parse_results(). no job_status is set to "error".

File: arc/job/adapters/ase_adapter.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+
+if TYPE_CHECKING:
+    from arc.level import Level
+    from arc.species.species import ARCSpecies
+    from arc.reaction import ARCReaction
+
+logger = get_logger()
+
+# Default mapping if not yet fully defined in settings.py
+DEFAULT_ASE_ENV = {
```
Comment (alongd) [inline arc/job/adapters/ase_adapter.py:25]: but settings defines `ASE_CALCULATORS_ENV`. Either drop `DEFAULT_ASE_ENV` or align it with the settings dict?

File: arc/job/adapters/scripts/ase_script.py [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+import yaml
+import numpy as np
+
+from ase import Atoms
+from ase.constraints import FixInternals
+from ase.optimize import BFGS, LBFGS, GPMin
+from ase.optimize.sciopt import SciPyFminBFGS, SciPyFminCG
+from ase.vibrations import Vibrations
+
+# Constants matched to ASE internal units (3.23.0+) for exact numerical matching
```
Comment (alongd) [inline arc/job/adapters/scripts/ase_script.py:22]: constants should go here: https://github.com/ReactionMechanismGenerator/ARC/blob/main/arc/constants.py
please consolidate

File: arc/job/adapters/scripts/ase_script.py [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+            opt_class = Sella
+        else:
+            opt_class = engine_dict.get(engine_name, BFGS)
+        opt = opt_class(atoms, logfile=os.path.join(os.path.dirname(input_path), 'opt.log'))
+        
+        try:
+            opt.run(fmax=fmax, steps=steps)
+            save_current_geometry(output, atoms, xyz)
+            output['sp'] = to_hartree(atoms.get_potential_energy())
+        except Exception as exc:
```
[the remark this answers is not in the document]
Comment (alongd) [reply arc/job/adapters/scripts/ase_script.py]: did you see this comment?

File: arc/job/adapters/scripts/ase_script.py [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+            from sella import Sella
+            opt_class = Sella
+        else:
+            opt_class = engine_dict.get(engine_name, BFGS)
+        opt = opt_class(atoms, logfile=os.path.join(os.path.dirname(input_path), 'opt.log'))
+        
+        try:
+            opt.run(fmax=fmax, steps=steps)
+            save_current_geometry(output, atoms, xyz)
+            output['sp'] = to_kJmol(atoms.get_potential_energy())
```
Comment (alongd) [inline arc/job/adapters/scripts/ase_script.py:355]: we should not store sp when the opt fails

File: arc/job/adapters/scripts/ase_script.py [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+    output = {}
+    
+    def save_current_geometry(out_dict, atoms_obj, input_xyz):
+        out_dict['opt_xyz'] = {
+            'coords': tuple(map(tuple, atoms_obj.get_positions().tolist())),
+            'symbols': input_xyz['symbols'],
+            'isotopes': input_xyz.get('isotopes') or tuple([None] * len(input_xyz['symbols']))
+        }
+
+    if job_type in ['sp', 'opt', 'conf_opt', 'freq', 'optfreq', 'directed_scan']:
```
Comment (alongd) [inline arc/job/adapters/scripts/ase_script.py]: for an opt job, this triggers an SCF/energy calc. then the opt runs and overwrites output['sp'].
This is minor, but we might multiply it by a huge number of calls for advanced work. 
So for any opt/optfreq/conf_opt/directed_scan job, that first energy evaluation is thrown away. i think that the first if should really be "sp" only

File: arc/job/adapters/scripts/ase_script.py [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+def get_calculator(calc_config: dict, charge: int = 0, multiplicity: int = 1):
+    """
+    Initialize the ASE calculator based on settings.
+    """
+    name = calc_config.get('calculator', '').lower()
+    kwargs = calc_config.get('calculator_kwargs', {})
+    
+    if name == 'torchani':
+        import torch
+        import torchani
```
Comment (alongd) [inline arc/job/adapters/scripts/ase_script.py:69]: unless there's a reason, please relocate all imports to the top of the file as a convention in ARC

File: arc/job/adapters/scripts/ase_script.py [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+
+
+def get_calculator(calc_config: dict, charge: int = 0, multiplicity: int = 1):
+    """
+    Initialize the ASE calculator based on settings.
+    """
+    name = calc_config.get('calculator', '').lower()
+    kwargs = calc_config.get('calculator_kwargs', {})
+    
+    if name == 'torchani':
```
Comment (alongd) [inline arc/job/adapters/scripts/ase_script.py:67]: The function accepts charge and multiplicity. xTB uses them, tut TorchANI MOPAC receive those same parameters and silently ignore them.
for TorchANI So ignoring is correct, just undocumented (should be logged)
But MOPAC should support charge/multiplicity ((via keywords like CHARGE= / DOUBLET etc.)

File: arc/job/adapters/ase_test.py [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+        self.assertEqual(self.job_1.force_constants, [1.0, 2.0, 3.0])
+        self.assertIsNotNone(self.job_1.xyz_out)
+        self.assertAlmostEqual(self.job_1.xyz_out['coords'][1][1], 0.76)
+
+    def test_to_kJmol(self):
+        """Test utility conversion function to_kJmol"""
+        self.assertAlmostEqual(to_kJmol(1.0), 96.48533, places=5)
+        self.assertAlmostEqual(to_kJmol(27.21138), 2625.499015202655, places=5)
+
+    def test_numpy_vibrational_analysis(self):
```
Comment (alongd) [inline arc/job/adapters/ase_test.py:147]: also test a >2 atom linear molecule

**Review by alongd — COMMENTED** [review]
Thanks @kfir4444! Please see a few more comments

File: arc/job/adapters/scripts/ase_script.py (hunk 1 above)
Comment (kfir4444) [reply arc/job/adapters/scripts/ase_script.py]: I think the number should by 5 anyway. Linear has 3N-5, bu we check the `abs(freqs[indices[i]])` anyways (3 lines down), and don't take anything smaller than 10, which all are supposed to be.

File: arc/job/adapters/ase_adapter.py (hunk 2 above)
Comment (kfir4444) [reply arc/job/adapters/ase_adapter.py:174]: Actually, no other adapter that I can find sets job_status to error.

File: arc/job/adapters/ase_adapter.py (hunk 3 above)
Comment (kfir4444) [reply arc/job/adapters/ase_adapter.py:25]: This is the default value. In `get_python_executable` we test it versus the `env_mapping = settings.get('ASE_CALCULATORS_ENV', DEFAULT_ASE_ENV)`. Can you explain what's wrong with the current implementation?

File: arc/job/adapters/scripts/ase_script.py (hunk 9 above)
Comment (kfir4444) [reply arc/job/adapters/scripts/ase_script.py:67]: I agree about the charge, but we don't have any way of interacting with the multiplicity. They way they do that in ASE is like this:
```python
        magmom = int(round(abs(atoms.get_initial_magnetic_moments().sum())))
        if magmom:
            s += (['DOUBLET', 'TRIPLET', 'QUARTET', 'QUINTET'][magmom - 1] +
                  ' UHF ')
```
So you must manually do `atoms.set_initial_magnetic_moments `, but this is unpredictable.
I think we can safely say that MOPAC is used for ground state species and throw a warning.
Generally, it seems like there is a lack of standardization in some of the calculators. E.g.- xTB (and the reccomended TBLite #899 ) interface uses `uhf`, so it would be difficult to generalize.

File: arc/job/adapters/scripts/ase_script.py (hunk 1 above)
Comment (kfir4444) [reply arc/job/adapters/scripts/ase_script.py]: 2nd thought, implemented it :smile: .

File: arc/job/adapters/scripts/ase_script.py (hunk 8 above)
Comment (kfir4444) [reply arc/job/adapters/scripts/ase_script.py:69]: There is a reason, you don't know up from which modules you should load, since the LOT is checked on runtime.

**Review by alongd — APPROVED** [review]
